file_util/src/file.cc

```diff
@@ -0,0 +1,135 @@
+#include <leatherman/file_util/file.hpp>
+#include <boost/nowide/fstream.hpp>
+#include <boost/nowide/cenv.hpp>
+#include <sstream>
+#include <leatherman/logging/logging.hpp>
+
+#include <wordexp.h>
+
+using namespace std;
+
+namespace leatherman { namespace file_util {
+
+    bool each_line(string const& path, function<bool(string&)> callback)
+    {
+        boost::nowide::ifstream in(path.c_str());
+        if (!in) {
+            return false;
+        }
+
+        string line;
+        while (getline(in, line)) {
+            if (!callback(line)) {
+                break;
+            }
+        }
+        return true;
+    }
+
+    string read(string const& path)
+    {
+        string contents;
+        if (!read(path, contents)) {
+            return {};
+        }
+        return contents;
+    }
+
+    bool read(string const& path, string& contents)
+    {
+        boost::nowide::ifstream in(path.c_str(), ios::in | ios::binary);
+        ostringstream buffer;
+        if (!in) {
+            return false;
+        }
+        buffer << in.rdbuf();
+        contents = buffer.str();
+        return true;
+    }
+
+    std::string shellExpand(std::string txt) {
+        // This will store the expansion outcome
+        wordexp_t result;
+
+        // Expand and check the success
+        if (wordexp(txt.c_str(), &result, 0) != 0 || result.we_wordc == 0) {
+            wordfree(&result);
+            return "";
+        }
+
+        // Get the expanded text and free the memory
+        std::string expanded_txt { result.we_wordv[0] };
+        wordfree(&result);
+        return expanded_txt;
+    }
+
+    bool file_readable(const std::string &file_path) {
+        bool exists { false };
+        if (file_path.empty()) {
+            LOG_WARNING("file path is an empty string");
+        } else {
+            boost::nowide::ifstream file_stream { file_path };
+            exists = file_stream.good();
+            file_stream.close();
+        }
+        return exists;
+    }
+
+    void atomic_write_to_file(const std::string &text,
+                              const std::string &file_path,
+                              std::ios_base::openmode mode = ios::binary) {
+        boost::nowide::ofstream ofs;
+        std::string tmp_name = file_path + "~";
+        ofs.open(tmp_name, mode);
+        if (!ofs.is_open()) {
+            throw std::runtime_error { "failed to open " + file_path };
```

Comment:
I was having trouble getting filesystem_error to work, as a second argument it wanted an error_code, which is an enum defined in boost::filesystem, but it didn't want to compile, most likely because I didn't chance on the right syntax. Do you have any insight?